### DISCUSSION TITLE
Using ethernet enabled ESP32 boards with this board?

### DISCUSSION THREAD

**Discussion by Blackra1n5** [CircuitSetup/Expandable-6-Channel-ESP32-Energy-Meter#40] [Q&A] [answered]
I would prefer using ethernet instead of wifi for reliability.

Would the [WESP32](https://wesp32.com/hardware/) or [Olimex's ESP32-POE-ISO](https://www.olimex.com/Products/IoT/ESP32/ESP32-POE-ISO/resources/ESP32-POE-ISO-GPIO.png) be used with the 6 Channel Energy Meter?
And is there a specific pins I should be using?

I have linked the GPIO pinouts of the 2 boards.

Thank You

**Comment by presslab-us** [comment]
I'm somewhat interested in this too.  I would prefer Ethernet as the unit is very close to my main network switch.

Looking at the pinout of both of those boards, it seems like neither would be a drop-in.  I guess it could be sky-wired or an adapter board made.

**Comment by CircuitSetup** [comment, marked as answer]
@Blackra1n5 neither of those would directly plug into the meter since the pins are different. You could wire up the SPI pins and power with jumpers, but that would be messy unless you made some kind of pcb adapter.

This looks like it could work though: https://www.electrodragon.com/product/lan8720-ethernet-breakout-board-phy/  but you'd have to solder on different headers to plug into the meter and the esp32.

**Comment by presslab-us** [comment]
Looking at the code here:
https://github.com/espressif/arduino-esp32/blob/0f772270fbb32099cad72053755d6b4c7e8e1f82/tools/sdk/include/ethernet/eth_phy/phy.h#L34

I think the RMII pins will conflict with the SPI pins.

**Comment by robotsflo** [comment]
The ATM90E32AS can be used with SPI or UART, the UART can be the solution with Ethernet LAN8720 ?

**Comment by CircuitSetup** [comment]
> Looking at the code here:
> https://github.com/espressif/arduino-esp32/blob/0f772270fbb32099cad72053755d6b4c7e8e1f82/tools/sdk/include/ethernet/eth_phy/phy.h#L34
> 
> I think the RMII pins will conflict with the SPI pins.

Yes, it would, since SPI uses 19 for MISO, which could be changed if you don't mind soldering a wire on the pcb, and changing it in software.

**Comment by CircuitSetup** [comment]
> The ATM90E32AS can be used with SPI or UART, the UART can be the solution with Ethernet LAN8720 ?

It wouldn't matter since the pcb is already setup for SPI. Either way you you would have to solder in a new wire or 2.

**Comment by robotsflo** [comment]
> 
> 
> > Looking at the code here:
> > https://github.com/espressif/arduino-esp32/blob/0f772270fbb32099cad72053755d6b4c7e8e1f82/tools/sdk/include/ethernet/eth_phy/phy.h#L34
> > I think the RMII pins will conflict with the SPI pins.
> 
> Yes, it would, since SPI uses 19 for MISO, which could be changed if you don't mind soldering a wire on the pcb, and changing it in software.

Would you have some shematics and code of this tweak ? I think this would be valuable for people will be interested by ethernet feature.

**Comment by CircuitSetup** [comment]
> Would you have some shematics and code of this tweak ? I think this would be valuable for people will be interested by ethernet feature.

Sorry I didn't reply to your question earlier. Honestly, I wouldn't recommend doing this just yet. I've been researching this for a new meter, and there are several things that have to work correctly. 

If you can wait until the new meter (it's actually being made for 3-phase power, but can be configured for split single phase), it will have a built in ethernet jack + an on board ESP32. I don't have an ETA, but I am going to start testing next week.

**Reply by missmah** [reply]
Any ETA on this? Did you abandon it? I just ordered two of the existing boards but this sounds like it would fit my needs much better!

**Comment by presslab-us** [comment]
I have successfully added a DM9051 Ethernet MAC/PHY.  It needs 5 pins, I have it attached to the HSPI port.  There were some challenges in software, it would be nice if the Arduino library was no longer needed.  But it seems to be working ok.

Using the 5 pins, it will reduce the number of boards that are able to be stacked.  But I can think of a way to reduce the number of GPIO needed for the CS signals.  With some simple logic on each add-on board there could be one CS for each board and another GPIO that toggles between the two ICs.

**Reply by CircuitSetup** [reply]
Do you have a schematic of how you set this up, and what you did in software? Last time I tried to use both SPI interfaces on the ESP32 it didn't work very well. 

Also, 5 pins on the DM9051 seems much better than 10 (including a PHY enable) needed for the LAN8720

**Reply by presslab-us** [reply]
The pinout is as follows:
```
DM9051 ESP32

SS     GPIO15
MOSI   GPIO13
MISO   GPIO12
CLK    GPIO14
INT    GPIO4
```

For the software I needed to integrate the espidf framework together with the arduino framework.  It would be better if the arduino framework could be removed altogether but this will require a lot of work.

There also seems to be some instabilities in the DM9051 driver which I did find a workaround for.

**Comment by CeeWeasel** [comment]
Not to bring up an old post, but does this mean I could take an extra [ethernet enabled QuniLED-ESP32](https://quinled.info/quinled-esp32-ethernet/), and use it to run a board, as long as I powered it separately, and configured it to use pins that aren't used by the LAN8720A built into the board (listed below)? I realize I'd have to jumper between the boards and it could get ugly. Right now, I'm just wondering if it is even possible.

ESP32 GPIO	FUNCTION
GPIO_5	        Power IC Enable trigger (3.3v High for On)
GPIO17	        Clock Signal (from ESP32)
GPIO18	        Ethernet (PHY_MDIO)
GPIO19	        Ethernet (TXD[0])
GPIO21	        Ethernet (TX_EN)
GPIO22	        Ethernet (TXD[1])
GPIO23	        Ethernet (PHY_MDC)
GPIO25	        Ethernet (RXD[0])
GPIO26	        Ethernet (RXD[1])
GPIO27	        Ethernet (CRS_DV)

**Reply by CircuitSetup** [reply]
That could work, as long as it still has 2 free GPIOs for cs and 18, 19, 23 for spi pins. You'd also either have to make a little wire harness, or adapter pcb. 

If it does meet those requirements, then I may have to get a hold of one to test.

**Comment by CeeWeasel** [comment]
I'll see what I can come up with. I just think it would be nice to use
that, along with a POE to USB extractor, to connect and power the board
over POE (which works just fine on the QuinLed-ESP32), and leave the whole
setup in the box or in the wall, without worrying about wifi connectivity.

**Reply by CircuitSetup** [reply]
If you power it via USB/POE only you wouldn't get the proper voltage reference. If you want to have both I would recommend removing the rectifier (square thing next to the voltage input).

BTW, I believe it has just enough pins for SPI, but you shouldn't use 12 (will not boot if pulled high), 34, or 35 (input only).

**Comment by vingenuity** [comment]
Has there been any progress with running these boards over ethernet?

I am interested in this, but I have an outdoor breaker box far from my router and I doubt the ESP32 will be able to reach it via wifi...

**Comment by CeeWeasel** [comment]
I haven't tested further. I went down the path of having an ESP32 with a
detachable/external antenna. That let me maintain wifi coverage over an
extended distance, without issues.

**Reply by vingenuity** [reply]
What ESP32 did you end up using? Was it drop-in?

**Comment by CeeWeasel** [comment]
It was a Nodemcu 32s with an external antenna. Something like this:
ESP32-DevKitC
core Board for ESP32 Development Board ESP32-WROOM-32U
<https://www.amazon.com/MELIFE-ESP32-DevKitC-Development-ESP32-WROOM-32U-Arduino/dp/B0811KLGDD/ref=sr_1_1?crid=1N2PJX7E2HAHY&dib=eyJ2IjoiMSJ9.nvqz3D12AHHICzTS50qHW3wPiggLXhfzDC9gc2AaKu2WPYzRgbAGncIf53iRoyW2YAnRlucdtPy0WwQpDH15jbdNzrE5mSc6LAQI9TmzPmQ6G_5_eEJa2Tn_8r4udcGBj6E9sUJT44ZZlTHQwFUq1oxTsDffhnGv-q0NVPYikjspn3MotV6XhL6JG_TJhrneXjbrun7OLlML9LPCfAa_hPz21yyszDbuzYUqIrpjnBv5UC4rvr3tps-5_a4nLhEGnElPOFRIEIGfR5TiiZi_u6c5PrlHZAa5AHqrF_xPWqU.SiZ8dw2oTNPqy8RfnaHpS_H1pkhR61BKPXfGD2VC70I&dib_tag=se&keywords=nodemcu%2Besp32%2Bexternal%2Bantenna&qid=1717418662&s=electronics&sprefix=nodemcu%2Besp32%2Bexternal%2Bantenna%2Celectronics%2C91&sr=1-1&th=1>
for
the board, along with a compatible antenna.

**Reply by vingenuity** [reply]
Thank you for the info!

Ordered all of the parts, and hopefully the external antenna can get me the range I need...

**Comment by TRSx80** [comment]
I have no idea how this compares, just something I came across during my recent research and thought I would share:

[WT32-ETH01 With ESPHome Driving “CircuitSetup Expandable 6 Channel ESP32 Energy Meter Main Board”](https://community.home-assistant.io/t/wt32-eth01-with-esphome-driving-circuitsetup-expandable-6-channel-esp32-energy-meter-main-board/709027)

BTW, really cool project you have here, I am sure I will be picking up some kit soon (just a few details to finish researching first)!

**Comment by CircuitSetup** [comment]
FYI, there is now an adapter board for the Lilygo T-ETH Lite ESP32S3 here: https://circuitsetup.us/product/6-channel-esp32-energy-meter-ethernet-adapter/?v=7516fd43adaa
